fuel-txpool/src/containers/dependency.rs

```diff
@@ -512,15 +513,15 @@ impl Dependency {
                     }
                 }
                 Input::Contract { contract_id, .. } => {
-                    // remove from contract
+                    // Input contract cases
+                    // 1. contract state no longer exists because the parent tx that created the
+                    //    contract was already removed from the graph
+                    // 2. contract state exists and this tx needs to be removed as a user of it.
+                    // 2.a. contract state can be removed if this is the last tx to use it.
                     let mut rem_contract = false;
-                    {
-                        let state = self
-                            .contracts
-                            .get_mut(contract_id)
-                            .expect("Contract should be present");
+                    if let Some(state) = self.contracts.get_mut(contract_id) {
                         state.used_by.remove(&tx.id());
-                        // if contract list is empty, remove contract from dependency.
+                        // if contract list is empty, flag contract state for removal.
                         if state.used_by.is_empty() {
```

Comment:
while looking at it, this check is missing  a `&& state.depth == 0` as only contracts loaded from db need to be remove from dependencies in this stage.

Reply:
good catch, added this case and also refactored to use a helper `is_in_database()` to make the code more readable.